omp/test/factorization/par_ilu_kernels.cpp

```diff
@@ -0,0 +1,192 @@
+        ref, gko::lend(csr_Omp), &l_nnz_Omp, &u_nnz_Omp);
+
+    ASSERT_EQ(l_nnz_Omp, l_nnz_Ref);
+    ASSERT_EQ(u_nnz_Omp, u_nnz_Ref);
+}
+
+
+TEST_F(ParIlu, KernelInitializeParILUIsEquivalentToRef)
+{
+    gko::size_type l_nnz_Ref{};
+    gko::size_type u_nnz_Ref{};
+    gko::size_type l_nnz_Omp{};
+    gko::size_type u_nnz_Omp{};
+
+    gko::kernels::reference::par_ilu_factorization::compute_nnz_l_u(
+        ref, gko::lend(csr_Ref), &l_nnz_Ref, &u_nnz_Ref);
+
+    gko::kernels::omp::par_ilu_factorization::compute_nnz_l_u(
+        ref, gko::lend(csr_Omp), &l_nnz_Omp, &u_nnz_Omp);
+
+    auto l_Ref = Csr::create(ref, csr_Ref->get_size(), l_nnz_Ref);
+    auto u_Ref = Csr::create(ref, csr_Ref->get_size(), u_nnz_Ref);
+
+    auto l_Omp = Csr::create(ref, csr_Omp->get_size(), l_nnz_Omp);
+    auto u_Omp = Csr::create(ref, csr_Omp->get_size(), u_nnz_Omp);
+
+    gko::kernels::reference::par_ilu_factorization::initialize_l_u(
+        ref, gko::lend(csr_Ref), gko::lend(l_Ref), gko::lend(u_Ref));
+    gko::kernels::omp::par_ilu_factorization::initialize_l_u(
+        ref, gko::lend(csr_Omp), gko::lend(l_Omp), gko::lend(u_Omp));
+
+    GKO_ASSERT_MTX_NEAR(l_Ref, l_Omp, 1e-14);
+    GKO_ASSERT_MTX_NEAR(u_Ref, u_Omp, 1e-14);
+}
+
+
+TEST_F(ParIlu, KernelComputeParILUIsEquivalentToRef)
+{
+    gko::size_type l_nnz_Ref{};
+    gko::size_type u_nnz_Ref{};
+    gko::size_type l_nnz_Omp{};
+    gko::size_type u_nnz_Omp{};
+
+    gko::size_type iterations{};
+
+    gko::kernels::reference::par_ilu_factorization::compute_nnz_l_u(
+        ref, gko::lend(csr_Ref), &l_nnz_Ref, &u_nnz_Ref);
+
+    gko::kernels::omp::par_ilu_factorization::compute_nnz_l_u(
+        ref, gko::lend(csr_Omp), &l_nnz_Omp, &u_nnz_Omp);
+
+    auto l_Ref = Csr::create(ref, csr_Ref->get_size(), l_nnz_Ref);
+    auto u_Ref = Csr::create(ref, csr_Ref->get_size(), u_nnz_Ref);
+
+    auto l_Omp = Csr::create(ref, csr_Omp->get_size(), l_nnz_Omp);
+    auto u_Omp = Csr::create(ref, csr_Omp->get_size(), u_nnz_Omp);
+
+    gko::kernels::reference::par_ilu_factorization::initialize_l_u(
+        ref, gko::lend(csr_Ref), gko::lend(l_Ref), gko::lend(u_Ref));
+    gko::kernels::omp::par_ilu_factorization::initialize_l_u(
+        ref, gko::lend(csr_Omp), gko::lend(l_Omp), gko::lend(u_Omp));
+
+    auto coo_Ref = Coo::create(ref);
+    csr_Ref->convert_to(gko::lend(coo_Ref));
+
+    auto u_transpose_Ref = u_Ref->transpose();
+
+    gko::kernels::reference::par_ilu_factorization::compute_l_u_factors(
+        ref, iterations, gko::lend(coo_Ref), gko::lend(l_Ref),
+        gko::lend(u_Ref));
+
+    auto coo_Omp = Coo::create(omp);
+    csr_Omp->convert_to(gko::lend(coo_Omp));
+
+    auto u_transpose_Omp = u_Omp->transpose();
+
+    gko::kernels::omp::par_ilu_factorization::compute_l_u_factors(
+        omp, iterations, gko::lend(coo_Omp), gko::lend(l_Omp),
+        gko::lend(u_Omp));
+
+    GKO_ASSERT_MTX_NEAR(l_Ref, l_Omp, .3);
+    GKO_ASSERT_MTX_NEAR(u_Ref, u_Omp, .3);
```

Comment:
For the 3 tests, I would recommend creating 3 functions that generate the data that you need for a given executor:
- The first one simply calls the `compute_nnz_l_u` kernel and returns (or writes in given pointers) the nnz count.
- The second one calling the first, followed by creating and initializing the matrices (which are returned)
- The third one calling the second one, followed by calling `compute_l_u_factors`

That way, in the tests, you can simply call these functions and compare the output. That has three benefits:
1. Since you reuse the code, there is less chance of a copy paste error (not that I have seen one)
2. You have a lot less code duplication, and the test gets a lot easier to understand / read
3. You can use the same function for both the `Omp` and the `Reference` executor.